[WIP] Support for Windows ( bfirsh/whalebrew#29 )

3846masa opened the pull request:
Whalebrew on Windows 🐋 🎉 🐳  ( bfirsh/whalebrew#29 )

I checked on cmd and MSYS2 bash with Docker for Windows.

---

# PR for Windows

## syscall -> exec (https://github.com/3846masa/whalebrew/commit/8a6d78379dbc0357fa95679ce10c8f5db24d8301)
To change `syscall.Exec` to `exec.Command` because `syscall.Exec` causes Error on Windows.

## ~Fix checking args~
~First argument checking hack is not working on Windows. I changed it.~

## Add batch file support (https://github.com/3846masa/whalebrew/commit/ffbdd151865b700478a742c31f1af16d783666ed)
When `whalebrew install` on Windows, generate batch file.
~Batch file passes shell file and args to `whalebrew`.~

## Fix default WHALEBREW_INSTALL_PATH (https://github.com/3846masa/whalebrew/commit/74db526e3efce7250e19cc1d85683e394034fa69)
I have no idea where is good for install_path on Windows.
Temporally, `WHALEBREW_INSTALL_PATH` is `C:\whalebrew` if on Windows.

### Add install script for Windows (https://github.com/3846masa/whalebrew/commit/94a625d808d9fb4d9f11ea996e70bc427026c926)
Install script will download `whalebrew.exe` and set `Path` and `WHALEBREW_INSTALL_PATH`.

## `path.Join` -> `filepath.Join` (https://github.com/3846masa/whalebrew/commit/7453d3797a457968005073a74a502dda3e9638ca)
c.f. ) http://stackoverflow.com/questions/9371031/how-do-i-create-crossplatform-file-paths-in-go

## Rewrite test for Windows (https://github.com/3846masa/whalebrew/commit/bc12aa0da1b47f5d6f0a608c348423ad6ca50767)

### Add `MakePackagePath` function (https://github.com/3846masa/whalebrew/commit/fe06f35661c97bd768341767ae9e9e2b71d252f4)
Because package file name is different on Windows and *nix

---

# TODO

- [x] CI Test for Windows
- [x] Best solution for `WHALEBREW_INSTALL_PATH` on Windows

---

This is first time to write golang. Please check changes. 🙏 
Thanks.

Foucl commented:
I can confirm that this works on Win 10 x64, Preview 15019 with latest docker for windows beta release. I used it with cmd, but also have msys2 installed and its binaries in my PATH (don't know if that's relevant).
tiny note: maybe create the install directory if it doesn't exist (`whalebrew install` currently throws an error if it doesn't).

bfirsh commented on main.go:
```diff
@@ -3,17 +3,20 @@ package main
+		// Check if not command exists
+		if _, _, err := cmd.RootCmd.Find(os.Args); err != nil {
+			// Check if file exists
+			if _, err := os.Stat(os.Args[1]); err == nil {
```
This seems even more flimsy than my hack. ;) e.g. I can envisage lots of stuff unexpectedly not working if they collide with command names.

You could just run `whalebrew run [pkg]` in the batch script.

bfirsh commented:
Thanks for this @3846masa! This is a great start.

Some thoughts:
* syscall.Exec can't be trivially replaced (exit code won't be returned, signals won't be passed, etc. Probably lots of unknown unknowns here.)
* I think we can come up with a neater solution than writing two files. I did the YAML file with shebang method because it works neatly on *nix, but that's clearly not the case on Windows.
* I don't run windows, and I'm sure a lot of contributors won't, so we need thorough integration tests and CI to make sure we don't break it. This is the main reason I want to be conservative about including this - it'll slow down development.
* Presumably we'll need some different installation instructions for windows, including setting up paths.

mapitman commented:
Didn't realize this PR existed and I've been working on my own branch to get whalebrew working on Windows. I found a problem with using `exec.Command` and hooking up stdin, stdout and stderr. I tested with the whalebrew/lynx package and it doesn't work so well. Packages that don't require any input other than what's passed on the command line seem to work fine. Maybe whalebrew can just write out the full Docker command into the batch files?

mapitman commented:
Even running the `whalebrew/lynx` docker container directly from the commandline doesn't seem to work very well on Windows. It's not as bad as when using `exec.Command` from inside whalebrew, but still not very usable.

mapitman commented:
More info: when you pass a commandline argument to the lynx command, it hangs and I thought it was only on windows, but it is happening on macOS as well:

`lynx http://google.com`

So, I don't think there is a problem using command.Exec on Windows. I figured out a way to write out the batch files with YAML embedded too.

3846masa commented:
Thanks for comments. I changed code.

- Use `exec.Command` on **only Windows**
  - Fix to return status code of `exec.Command`
- Batch file includes YAML (This is valid as YAML and batch file)
  - This solution generates only one file
  ```batch
  :: |
    @( whalebrew run %~f0 %* || exit /b %ERRORLEVEL% ) && exit /b 0
  image: whalebrew/whalesay
  ```
- Revert args hack
  - I agree @bfirsh opinion.
  - For windows, user just run `whalebrew run [pkg]` via batch file

3846masa commented:
I wrote TODO in [first comment](https://github.com/bfirsh/whalebrew/pull/34#issue-204388665) and change PR title.

3846masa commented:
I changed.

- Merge v0.0.4 and Fix for Windows
  - 3846masa/whalebrew@bb652d4b3227bb4b14bde38f751328db3bc7e5af
  - 3846masa/whalebrew@8078165d1037f1db29361a2cfd089cfd79dbef62
  - 3846masa/whalebrew@8746496abc600c3f8bc3194d9d34636ec71d100c
- `path.Join` -> `filepath.Join`
  - 3846masa/whalebrew@7453d3797a457968005073a74a502dda3e9638ca
- **Rewrite test**
  - See results via [Appveyor](https://ci.appveyor.com/project/3846masa/whalebrew/build/7) and [Travis CI](https://travis-ci.org/3846masa/whalebrew/builds/201123586).
  - 3846masa/whalebrew@bc12aa0da1b47f5d6f0a608c348423ad6ca50767
  - Add `MakePackagePath` function
    - Because package file name is different on Windows and *nix
    - 3846masa/whalebrew@fe06f35661c97bd768341767ae9e9e2b71d252f4
  - Add appveyor.yml
    - Appveyor is CI for Windows
    - 3846masa/whalebrew@a0e486cbf6ebf2cc7056753343c88b1561f4c98e
  - Fix .travis.yml for forked project
    - 3846masa/whalebrew@21bcf1436a7c5afa0613da8c7e5cdfabf166cc0f

3846masa commented:
Bump to 0.1.0.
Also add [**install script for Windows**](https://github.com/3846masa/whalebrew/blob/support-windows-rebase/install.ps1).

bfirsh commented:
Nice, thank you @3846masa! (and @mapitman!) I will take a proper look at this when I get some time.

Would you be willing to help maintain the Windows support? I don't have a Windows machine so I am worried it will deteriorate unless somebody's keeping an eye on it. :)

Whalebrew seems pretty stable and feature complete now, so I doubt there'll be much involved.

3846masa commented:
Ok, I'll help maintain for Windows.
I want whalebrew on Windows because I use Windows machine everyday 👍

3846masa committed:
Support for Windows ( bfirsh/whalebrew#29 )

Squashed commit of the following:

commit b677dbc22021aac0680369376af46e30b8aeae32
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Sat Oct 28 11:35:30 2017 +0900

    Fix golang version

commit 43bc3636d2cf9d9cc8838fbea781967e6538b525
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Sat Oct 28 11:12:42 2017 +0900

    Fix shebang for Windows

commit ed225dfea8b2315ec6e045c347e28fc41c4f25ca
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Fri Sep 8 10:29:39 2017 +0900

    Fix CI

commit 94a625d808d9fb4d9f11ea996e70bc427026c926
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Fri Sep 8 10:26:49 2017 +0900

    Fix windows installer

commit 0bea8fbe807f98f9405df227a089459913ef7704
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Fri Sep 8 10:26:20 2017 +0900

    Fix README for Windows installer

commit 237a4a71da38e117c7d0fbfb22b0c01f5866b32a
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 19:11:19 2017 +0900

    Fix build for Windows

commit bc508d787bf0cd8f377bad01795f0fc1a601722c
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 18:59:16 2017 +0900

    Fix to set API version (moby/moby#32779)

commit 442be2f65b6e13ea0ef33234293b4de77b450efb
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 18:03:01 2017 +0900

    Update README for Windows

commit e83b2ed21a58e849570652234561c32f35623a32
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 18:01:01 2017 +0900

    Add installer for Windows

commit 776b931817d58876695ab2a01273b675a5498b5a
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 18:00:38 2017 +0900

    Build for Windows

commit bc01aef06449423b868bfe6a4525a39548319035
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 17:43:43 2017 +0900

    Fix `run.go` for Windows

commit 63b94903e3609a8c862d80cc94cc305a08579f1b
Merge: a72237f 0fea7e2
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 17:43:01 2017 +0900

    Merge tag '0.1.0' of https://github.com/bfirsh/whalebrew into support-windows

commit a72237fde18e08cf673960625a90532c04997f0e
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 17:26:29 2017 +0900

    Fix `ForceInstall` for windows

commit de458d9ba5f2c59ad9d573e47fe46f59cd93a8c7
Merge: 03d00bf e7d0792
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Thu Sep 7 17:22:41 2017 +0900

    Merge tag '0.0.5' into support-windows

commit 03d00bf7fe975021b629ff9a9dcc5d49832bed7c
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Tue Feb 14 00:01:52 2017 +0900

    Fix batch file's shebang

commit 21bcf1436a7c5afa0613da8c7e5cdfabf166cc0f
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 22:19:25 2017 +0900

    Fix .travis.yml for forked project

commit a0e486cbf6ebf2cc7056753343c88b1561f4c98e
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 21:59:19 2017 +0900

    Add appveyor.yml

commit fe06f35661c97bd768341767ae9e9e2b71d252f4
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 21:34:46 2017 +0900

    MakePackagePath

commit bc12aa0da1b47f5d6f0a608c348423ad6ca50767
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 21:23:31 2017 +0900

    Rewrite test

commit 7453d3797a457968005073a74a502dda3e9638ca
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 20:28:33 2017 +0900

    `path.Join` -> `filepath.Join`

commit 4ded1a223285f072a901925fc21a63de68410eef
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 16:53:51 2017 +0900

    LF -> CRLF in batch file

commit 8078165d1037f1db29361a2cfd089cfd79dbef62
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 16:34:08 2017 +0900

    Fix edit for Windows

commit 8746496abc600c3f8bc3194d9d34636ec71d100c
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 16:15:55 2017 +0900

    Fix run for windows

commit bb652d4b3227bb4b14bde38f751328db3bc7e5af
Merge: 8a6d783 e7e6427
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Mon Feb 13 16:08:42 2017 +0900

    Merge v0.0.4

    Merge remote-tracking branch 'origin/master' into support-windows

commit 8a6d78379dbc0357fa95679ce10c8f5db24d8301
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Sun Feb 5 21:11:59 2017 +0900

    Support to run and return status code on Windows

commit ce3d298aef065c9b4f34943daad7b45e9eb6e03f
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Sun Feb 5 20:58:21 2017 +0900

    Revert "syscall -> exec (for working on windows)"

    This reverts commit 276a9cd82b77b7280d8ebe3f362da7c3b7e40196.

commit ffbdd151865b700478a742c31f1af16d783666ed
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Sun Feb 5 20:50:08 2017 +0900

    Add batch file support (for working on windows)

commit cdc5a05c2ec6f6962eb0e05edc2f0c453ecbd28a
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Sun Feb 5 20:34:30 2017 +0900

    Revert "Add batch file support (for working on windows)"

    This reverts commit aba79b0c5a76aca84d29b2706c62bc5142f831e6.

commit 10e10a0d172706241d350bfbfe9ac027006dbd86
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Fri Feb 3 00:53:23 2017 +0900

    Revert "Fix checking args (for working on windows)"

    This reverts commit 926f422e1dffdc35444e74bde2f52279a65cb3c3.

commit 74db526e3efce7250e19cc1d85683e394034fa69
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Wed Feb 1 03:14:59 2017 +0900

    Fix default WHALEBREW_INSTALL_PATH (for working on windows)

commit aba79b0c5a76aca84d29b2706c62bc5142f831e6
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Wed Feb 1 03:06:00 2017 +0900

    Add batch file support (for working on windows)

commit 926f422e1dffdc35444e74bde2f52279a65cb3c3
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Wed Feb 1 01:45:51 2017 +0900

    Fix checking args (for working on windows)

commit 276a9cd82b77b7280d8ebe3f362da7c3b7e40196
Author: 3846masa <3846masahiro+git@gmail.com>
Date:   Wed Feb 1 01:36:56 2017 +0900

    syscall -> exec (for working on windows)

asadakbarml commented:
Is there a timeline for this?

tjamet commented:
Hi @asadakbarml thanks for the heads up. 
There is currently no work being done on windows.
I would be happy though to accept pull requests to do so.
